core/chain-spec/src/extension.rs

```diff
@@ -0,0 +1,363 @@
+		Self { base, forks }
+	}
+
+	/// Return a set of parameters for `Group` including all forks up to `block` (inclusive).
+	pub fn at_block(&self, block: B) -> T {
+		let mut start = self.base.clone().to_fork();
+
+		for (_, fork) in self.forks.range(..=block) {
+			start.combine_with(fork.clone());
+		}
+
+		start
+			.to_base()
+			.expect("We start from the `base` object, so it's always fully initialized; qed")
+	}
+}
+
+impl<B, T> IsForks for Forks<B, T> where
+	B: Ord + 'static,
+	T: Group + 'static,
+{
+	type BlockNumber = B;
+	type Extension = T;
+}
+
+impl<B: Ord + Clone, T: Group + Extension> Forks<B, T> where
+	T::Fork: Extension,
+{
+	/// Get forks definition for a subset of this extension.
+	///
+	/// Returns the `Forks` struct, but limited to a particular type
+	/// within the extension.
+	pub fn for_type<X>(&self) -> Option<Forks<B, X>> where
+		X: Group + 'static,
+	{
+		let base = self.base.get::<X>()?.clone();
+		let forks = self.forks.iter().filter_map(|(k, v)| {
+			Some((k.clone(), v.get::<Option<X::Fork>>()?.clone()?))
+		}).collect();
+
+		Some(Forks {
+			base,
+			forks,
+		})
+	}
+}
+
+impl<B, E> Extension for Forks<B, E> where
+	B: Serialize + DeserializeOwned + Ord + Clone + 'static,
+	E: Extension + Group + 'static,
+{
+	type Forks = Self;
+
+	fn get<T: 'static>(&self) -> Option<&T> {
+		use std::any::{TypeId, Any};
+
+		match TypeId::of::<T>() {
+			x if x == TypeId::of::<E>() => Any::downcast_ref(&self.base),
+			_ => self.base.get(),
+		}
+	}
+
+	fn forks<BlockNumber, T>(&self) -> Option<Forks<BlockNumber, T>> where
+		BlockNumber: Ord + Clone + 'static,
+		T: Group + 'static,
+		<Self::Forks as IsForks>::Extension: Extension,
+		<<Self::Forks as IsForks>::Extension as Group>::Fork: Extension,
+	{
+		use std::any::{TypeId, Any};
+
+		if TypeId::of::<BlockNumber>() == TypeId::of::<B>() {
+			Any::downcast_ref(&self.for_type::<T>()?).cloned()
+		} else {
+			self.get::<Forks<BlockNumber, <Self::Forks as IsForks>::Extension>>()?
+				.for_type()
+		}
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use chain_spec_derive::{ChainSpecGroup, ChainSpecExtension};
+	// Make the proc macro work for tests and doc tests.
+	use crate as substrate_chain_spec;
+
+	#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, ChainSpecGroup)]
+	#[serde(deny_unknown_fields)]
+	pub struct Extension1 {
+		pub test: u64,
+	}
+
+	#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, ChainSpecGroup)]
+	#[serde(deny_unknown_fields)]
+	pub struct Extension2 {
+		pub test: u8,
+	}
+
+	#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, ChainSpecGroup, ChainSpecExtension)]
+	#[serde(deny_unknown_fields)]
+	pub struct Extensions {
+		pub ext1: Extension1,
+		pub ext2: Extension2,
+	}
+
+	#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, ChainSpecExtension)]
+	#[serde(deny_unknown_fields)]
+	pub struct Ext2 {
+		#[serde(flatten)]
+		ext1: Extension1,
+		#[forks]
+		forkable: Forks<u64, Extensions>,
+	}
```

Comment:
I'm not sure I see the whole need for this derive macros, if it is only for constraining inputs this could be done with a simple function declared as such:
```rust
constrain! {
    non_forkable {
        ext1: Extension1,
    },
    forkable {
        ext1: Extension1,
        ext2: Extension2,
    },
}
```

With constrain declare as a proc macro or macro_rules, and it just check that input has one and only one field for each non_forkable and same for forkable, plus possible forks with at most one field for each forkable.

Do I miss some feature here ?

Reply:
`ChainSpecGroup` implements quite a lot of stuff, namely for struct:
```rust
struct Example {
 a: u64,
 b: u64,
}
```
it implements
```rust
struct ExampleFork {
 a: Option<u64>,
 b: Option<u64>,
}
```
and a bunch of traits to tie this too together. This is needed since you might want to hard fork only one of the parameters and leave previous one at previous values (see a test here: https://github.com/paritytech/substrate/blob/3171fc5063f9f32bb564da4076d42375fa2d03fb/core/chain-spec/src/extension.rs#L304)

The `ChainSpecExtension` is a bit different as it implements `Extension` trait, which has two methods:
- `get` which allows you to get a parameter of specific type (no mater how it's named in the struct)
- `forks` which allows to get `Forks` for specific type.
The `#[forks]` attribute is needed to figure out which field holds the forks.

But back to your proposal:
1. I think there is still a need to have `ChainSpecGroup` macro, to generate fork structs for `Extension1` and `Extension2`.
2. We still need to generate intermediate `struct Extension { ext1: Extension1, ext2: Extension2 }` to (de)serialize to JSON.
3. With current trait structure we still need to implement `Extension` trait for it.

So I find the derive macro a bit more readable than your proposal, unless we could get rid of some additional things (like some traits).

Reply:
oh Fork are used for nested structure as well. then ok it make sense.
Example could be a field of another structure here.